compiler/src/iree/compiler/Codegen/LLVMGPU/TransformDialectExtensions/TransformDialectLLVMGPUExtensions.cpp

```diff
@@ -359,6 +305,116 @@ static FailureOr<VectorDistributionResult> vectorDistribution(
   return VectorDistributionResult{warpOp};
 }
 
+// TODO: Refactor in a generic util that can be reused.
+static HAL::ExecutableExportOp getExecutableExportOpForFunc(
+    HAL::ExecutableVariantOp halExecutableVariantOp, func::FuncOp funcOp) {
+  if (!halExecutableVariantOp || !funcOp) return {};
+  HAL::ExecutableExportOp exportOp;
+  halExecutableVariantOp->walk([&](HAL::ExecutableExportOp op) {
+    if (op.sym_name() != funcOp.getName()) WalkResult::advance();
+    exportOp = op;
+    WalkResult::interrupt();
+  });
+  return exportOp;
+}
+
+FailureOr<vector::WarpExecuteOnLane0Op>
+transform_dialect::VectorWarpExecuteOnLane0Op::applyToOne(
+    scf::IfOp ifOp, transform::TransformState &state) {
+  assert(isa<HAL::ExecutableVariantOp>(state.getTopLevel()) &&
+         "requires HAL::ExecutableVariantOp toplevel so that IR is properly "
+         "isolated. This is required so we can safely inspect the "
+         "HAL::ExecutableExportOp under multi-threaded pass assumptions.");
+
+  auto halExecutableVariantOp =
+      ifOp->getParentOfType<HAL::ExecutableVariantOp>();
+  auto funcOp = ifOp->getParentOfType<func::FuncOp>();
+  HAL::ExecutableExportOp exportOp =
+      getExecutableExportOpForFunc(halExecutableVariantOp, funcOp);
+  assert(halExecutableVariantOp && funcOp && exportOp && "missing export op");
+
+  auto maybeAttr = exportOp.workgroup_size();
+  // TODO: Pervasive 3 constant in IREE.
+  if (!maybeAttr || maybeAttr->size() != 3)
+    return exportOp->emitError(
+        "export op must have workgroup_size attribute set with 3 entries");
+
+  int64_t workgroupSizeX = (*maybeAttr)[0].cast<IntegerAttr>().getInt();
+  int64_t warpSize = getWarpSize();
+  if (workgroupSizeX % warpSize != 0) {
+    exportOp->emitWarning()
+        << "vector distribution requires workgroup size for x to be a "
+        << "multiple of the warp size: " << workgroupSizeX << " vs " << warpSize
+        << " --- the transform is not applied";
+    return vector::WarpExecuteOnLane0Op();
```

Comment:
Should this rather return failure? Returning a null op sounds dangerous to me, it will get mapped to a handle, inspected, etc.

Reply:
This is intended, I added a comment.

Reply:
My gut sense still says it's dangerous. I won't block on this, but be aware of this any time you see a segfault. `TransformState` maintains a bidirectional mapping between payload IR ops and handles, so it will map the null operation to some handle (and if you get more than one such handle, it will assert). It is not unlikely that it accesses the payload operation without checking somewhere, I wrote that code with the assumption that null operation is an error and it never happens at that level because the error should have been either reported or silenced.

Reply:
ack, makes sense, thanks!